packages/tailwindcss/src/compat/container.ts

```diff
@@ -0,0 +1,112 @@
+import { atRule, decl, type AstNode } from '../ast'
+import type { DesignSystem } from '../design-system'
+import { compareBreakpoints } from '../utils/compare-breakpoints'
+import type { ResolvedConfig } from './config/types'
+
+export function registerContainerCompat(userConfig: ResolvedConfig, designSystem: DesignSystem) {
+  let container = userConfig.theme.container || {}
+
+  if (typeof container !== 'object' || container === null) {
+    return
+  }
+
+  let rules = buildCustomContainerUtilityRules(container, designSystem)
+
+  if (rules.length === 0) {
+    return
+  }
+
+  designSystem.utilities.static('container', () => structuredClone(rules))
+}
+
+export function buildCustomContainerUtilityRules(
+  {
+    center,
+    padding,
+    screens,
+  }: {
+    center?: boolean
+    padding?: string | {}
+    screens?: {}
+  },
+  designSystem: DesignSystem,
+): AstNode[] {
+  let rules = []
+  let breakpointOverwrites: null | Map<string, { nodes: AstNode[]; rule: string }> = null
+
+  if (center) {
+    rules.push(decl('margin-inline', 'auto'))
+  }
+
+  if (
+    typeof padding === 'string' ||
+    (typeof padding === 'object' && padding !== null && 'DEFAULT' in padding)
+  ) {
+    rules.push(
+      decl('padding-inline', typeof padding === 'string' ? padding : (padding.DEFAULT as string)),
+    )
+  }
+
+  if (typeof screens === 'object' && screens !== null) {
+    breakpointOverwrites = new Map()
+
+    // When setting a the `screens` in v3, you were overwriting the default
+    // screens config. To do this in v4, you have to manually unset all core
+    // screens.
+
+    let breakpoints = Array.from(designSystem.theme.namespace('--breakpoint').entries())
+    breakpoints.sort((a, z) => compareBreakpoints(a[1], z[1], 'asc'))
+    if (breakpoints.length > 0) {
+      let [key] = breakpoints[0]
+      // Unset all default breakpoints
+      rules.push(
+        atRule('@media', `(width >= theme(--breakpoint-${key}))`, [decl('max-width', 'none')]),
+      )
+    }
+
+    for (let [key, value] of Object.entries(screens)) {
+      // We're inlining the breakpoint values because the screens configured in
+      // the `container` option do not have to match the ones defined in the
+      // root `screen` setting.
+      breakpointOverwrites.set(key, {
+        rule: `(width >= ${value})`,
+        nodes: [decl('max-width', value)],
+      })
```

Comment:
Can we just use `atRule` here rather than an object that's turned into one further down?